[mod] limiter: reduce request rates for requests without a ping

return42 opened the pull request:
## What does this PR do?

Adds a very simple method to improve the limiter / *send a ping from the UI*.

## Why is this change important?

Too many bots pass the limiter and with their requests they worsen the acceptance of SearXNG engines.  In the issue Tracker we find many reports and comments about CAPTCHA and blocking of engines.  The requests of the engines can't be further improved to avoid CAPTCHA.

As long as bot requests are passed through from SearXNG, SearXNG will be classified as a bot by the origin search-engines.

The only way to improve this state is to develop methods that allow the limiter to better detect bots.

## Author's checklist

- We can add a token to the limiter URL
- This binds the limit of request rates to the `simple` theme.

return42 committed:
[mod] limiter: reduce request rates for requests without a ping

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] limiter: add random token to the limiter URL

By adding a random component in the limiter URL a bot can no longer send a ping
by request a static URL.

Related: https://github.com/searxng/searxng/pull/2357#issuecomment-1518525094
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

unixfox commented on searx/plugins/limiter.py:
```diff
@@ -40,6 +42,39 @@
+def ping():
+    redis_client = redisdb.client()
+    user_agent = request.headers.get('User-Agent', 'unknown')
+    x_forwarded_for = request.headers.get('X-Forwarded-For', '')
```
what happen when `X-Forwarded-For` is empty like for local usage with a request directly to searxng without passing through a reverse proxy?

return42 replied:
> what happen when `X-Forwarded-For` is empty like for local usage with a request directly to searxng without passing through a reverse proxy?

`X-Forwarded-For` is a prerequisite of the limiter.  If this value is empty, the requests are counted in the context of an empty IP (like any other IP). / This behavior is the same, with or without this PR.

return42 commented on searx/plugins/limiter.py:
```diff
@@ -40,6 +42,39 @@
+        token = token.decode('UTF-8')
+    else:
+        token = ''.join(random.choice(string.ascii_lowercase + string.digits) for _ in range(8))
+        redis_client.set(TOKEN_KEY, token, ex=600)
```
Here the token expires after 600 sec

return42 commented on searx/webapp.py:
```diff
@@ -648,6 +649,13 @@ def health():
     return Response('OK', mimetype='text/plain')
 
 
+@app.route('/limiter<token>.css', methods=['GET', 'POST'])
```
Here the token is expected in the URL / instead of an (empty) css we can return a (empty) html.

return42 commented on searx/templates/simple/base.html:
```diff
@@ -17,6 +17,9 @@
   <link rel="stylesheet" href="{{ url_for('static', filename='css/searxng.min.css') }}" type="text/css" media="screen" />
   {% endif %}
+  {% if get_setting('server.limiter') %}
+  <link rel="stylesheet" href="/limiter{{ limiter_token }}.css" type="text/css" media="screen" />
```
If css is not loaded by browsers like lynx .. will a `<iframe>` be loaded?

unixfox replied:
lynx doesn't automatically load the iframe nor it loads images nor javascript. the only way to detect that it's a terminal-based browser is to fingerprint the browser with methods like https://github.com/searxng/searxng/discussions/2264. as the user agent header is not trustworthy

return42 commented on searx/templates/simple/base.html:
```diff
@@ -17,6 +17,9 @@
   {% else %}
   <link rel="stylesheet" href="{{ url_for('static', filename='css/searxng.min.css') }}" type="text/css" media="screen" />
   {% endif %}
+  {% if get_setting('server.limiter') %}
```
> When redis is not configured,

The limiter request is only inserted if the limiter is activated.

return42 commented on searx/plugins/limiter.py:
```diff
@@ -40,6 +42,39 @@
+    x_forwarded_for = request.headers.get('X-Forwarded-For', '')
+
+    ping_key = PING_KEY + user_agent + x_forwarded_for
+    redis_client.set(secret_hash(ping_key), 1, ex=600)
```
>> But the URL of the limiter is changed every 10 min by a random token / compare my comments above.

> It doesn't really matter, if the user does multiple searches in a row then for the second time he visits the search page the "ping" CSS resource will be cached, and the issue may appear.

The token is also stored for 600 sec.  Only when the token changes after 10min, it must be "unlocked" again.

unixfox replied:
Mmh OK I understand the implementation. Then it's pretty fine in this case.

return42 commented on searx/plugins/limiter.py:
```diff
@@ -56,9 +91,20 @@ def is_accepted_request() -> bool:
+        if redis_client.get(secret_hash(ping_key)):
+            logger.debug('got a ping')
+            c_burst_max = 15
+            c_10min_max = 150
```
here the token, that have been stored (for 10min) is reused.

unixfox commented on searx/plugins/limiter.py:
```diff
@@ -56,9 +91,20 @@ def is_accepted_request() -> bool:
-        if c_burst > 15 or c_10min > 150:
+        if c_burst > c_burst_max or c_10min > c_10min_max:
             logger.debug("BLOCK %s: to many request", x_forwarded_for)
             return False
```
`BLOCK %s: to many request` => `BLOCK %s: too many requests`

return42 committed:
[mod] limiter -> botdetection: modularization and documentation

In order to be able to meet the outstanding requirements, the implementation is
modularized and supplemented with documentation.

This patch does not contain functional change, except it fixes issue #2455

----

Aktivate limiter in the settings.yml and simulate a bot request by::

    curl -H 'Accept-Language: de-DE,en-US;q=0.7,en;q=0.3' \
         -H 'Accept: text/html'
         -H 'User-Agent: xyz' \
         -H 'Accept-Encoding: gzip' \
         'http://127.0.0.1:8888/search?q=foo'

In the LOG:

    DEBUG   searx.botdetection.link_token : missing ping for this request: .....

Since ``BURST_MAX_SUSPICIOUS = 2`` you can repeat the query above two time
before you get a "Too Many Requests" response.

Closes: https://github.com/searxng/searxng/issues/2455
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] limiter: add config file /etc/searxng/limiter.toml

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] limiter.toml: botdetection.ip_limit turn off link_token by default

To activate the ``link_token`` method in the ``ip_limit`` method add the
following to your ``/etc/searxng/limiter.toml``::

   [botdetection.ip_limit]
   link_token = true

Related: https://github.com/searxng/searxng/pull/2357#issuecomment-1554116941
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] limiter: ip_limt - monitore suspicious IPs

To intercept bots that get their IPs from a range of IPs, there is a
``SUSPICIOUS_IP_WINDOW``.  In this window the suspicious IPs are stored for a
longer time.  IPs stored in this sliding window have a maximum of
``SUSPICIOUS_IP_MAX`` accesses before they are blocked.  As soon as the IP makes
a request that is not suspicious, the sliding window for this IP is droped.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/botdetection/link_token.py:
```diff
@@ -0,0 +1,126 @@
+from searx import redisdb
+from searx.redislib import secret_hash
+
+TOKEN_LIVE_TIME = 600
+"""Livetime (sec) of limiter's CSS token."""
```
> as the infinite scroll script won't make any request to the PING resource.

@unixfox: The ping is requested with the first query and is hold in the DB for 10min.

If the user stops scrolling for 10min and then continue he comes into trouble with the limiter.

To fix this, the ping can be renewed when a token is validated ...

    DEBUG   searx.botdetection.link_token : found ping for this request: 549553c3025039c76752a279925ab71522dbe3216171f6a8644b0e71c4a4f127

unixfox replied:
Yes it's a good idea to have the "Infinite scroll" feature renewing the ping resource.

return42 committed:
[mod] botdetection - improve ip_limit and link_token methods

- counting requests in LONG_WINDOW and BURST_WINDOW is not needed when the
  request is validated by the link_token method [1]

- renew a ping-key on validation [2], this is needed for infinite scrolling,
  where no new token (CSS) is loaded. / this does not fix the BURST_MAX issue in
  the vanilla limiter

- normalize the counter names of the ip_limit method to 'ip_limit.*'

- just integrate the ip_limit method straight forward in the limiter plugin /
  non intermediate code --> ip_limit now returns None or a werkzeug.Response
  object that can be passed by the plugin to the flask application / non
  intermediate code that returns a tuple

[1] https://github.com/searxng/searxng/pull/2357#issuecomment-1566113277
[2] https://github.com/searxng/searxng/pull/2357#discussion_r1208542206
[3] https://github.com/searxng/searxng/pull/2357#issuecomment-1566125979

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] correct determination of the IP for the request

For correct determination of the IP to the request the function
botdetection.get_real_ip() is implemented.  This fonction is used in the
ip_limit and link_token method of the botdetection and it is used in the
self_info plugin.

A documentation about the X-Forwarded-For header has been added.

[1] https://github.com/searxng/searxng/pull/2357#issuecomment-1566211059

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] limiter: replace real_ip by IPv4/v6 network

Closes: https://github.com/searxng/searxng/issues/2477
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] increase SUSPICIOUS_IP_WINDOW from one day to 30 days

In my tests I see bots rotating IPs (with endless IP lists).  If such a bot has
100 IPs and has three attempts (SUSPICIOUS_IP_MAX = 3) then it can successfully
send up to 300 requests in one day while rotating the IP.  To block the bots for
a longer period of time the SUSPICIOUS_IP_WINDOW, as the time period in which an
IP is observed, must be increased.

For normal WEB-browsers this is no problem, because the SUSPICIOUS_IP_WINDOW is
deleted as soon as the CSS with the token is loaded.

SUSPICIOUS_IP_WINDOW = 3600 * 24 * 30
  Time (sec) before sliding window for one suspicious IP expires.

SUSPICIOUS_IP_MAX = 3
  Maximum requests from one suspicious IP in the :py:obj:`SUSPICIOUS_IP_WINDOW`."""

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

unixfox commented on searx/botdetection/ip_limit.py:
```diff
@@ -0,0 +1,146 @@
+    # pylint: disable=too-many-return-statements
+    redis_client = redisdb.client()
+
+    if network.is_link_local and not cfg['botdetection.ip_limit.filter_link_local']:
```
Where is the list of link local addresses?

return42 replied:
The object `network` is a instance of class [IPv4Network](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Network) or [IPv6Network](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv6Network). These classes are python builtin types to handle IP addresses & networks.  The documentation:
-  [is_link_local](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_link_local) --> constant definition in the code for [IPv4](https://github.com/python/cpython/blob/77d25795862f19c6e3d647b76cfb10d5ce1f149c/Lib/ipaddress.py#L1542-L1570) and [IPv6](https://github.com/python/cpython/blob/77d25795862f19c6e3d647b76cfb10d5ce1f149c/Lib/ipaddress.py#L2262-L2292)

In the configuration [searx/botdetection/limiter.toml](https://github.com/searxng/searxng/pull/2357/commits/281e36f4b7848374535d5e953050ae73423191ca#diff-437a93b109d23719837e77f803998a188751dcc7f60cf711996939653defa967) link local addresses are unfiltered by default .. in a follow-up PR we can add an additional option for [is_private](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_private)

unixfox replied:
Shouldn't we offer the ability to add more link local networks? Some internal networks use public ranges, that's actually very common in IPv6 as there is usually no [NAT](https://en.wikipedia.org/wiki/Network_address_translation) involved and the "local" network is the public range of the network of computers behind the "modem" (internet connection).

return42 replied:
> Some internal networks use public ranges, 

Thanks for sharing your experience (I'm new to IPv6) .. very helpful :+1: 

When I added the option, I assumed that there would be many exceptions (such as private addresses or the cases you mentioned).  But link-local will always be link-local, that's irrevocable (right?).

With this knowledge in mind I decided to add the option link_local, other options I wanted to discuss in other PRs.

I see two possibilities for this PR:

a. we leave the link-local option as it is now.
b. we take it out of this PR again and move it to another PR where we discuss the question you asked "Shouldn't we offer the ability to add more link local networks?" in detail.

What do you think is the best way to proceed?

unixfox replied:
Yes I've used the word link local, but I actually meant private networks.

Maybe you should rename the variable and function names to "private_networks" if all the addresses located here are "whitelisted": https://github.com/python/cpython/blob/77d25795862f19c6e3d647b76cfb10d5ce1f149c/Lib/ipaddress.py#L2262-L2292

We can open another pull request for adding a support for more private networks.

return42 replied:
> Maybe you should rename the variable and function names to "private_networks" if all the addresses located here are "whitelisted":

Sorry, my link was not clear, "link local" is only:

- `IPv6Network('fe80::/10')`
- `IPv4Network('169.254.0.0/16')`

by this, the name `link_locale` is OK / I think.

> We can open another pull request for adding a support for more private networks.

:+1:

Do you see any other questionable in this PR?

unixfox replied:
Oh then the other IP addresses are not whitelisted?
```python
   _multicast_network = IPv6Network('ff00::/8')

    _private_networks = [
        IPv6Network('::1/128'),
        IPv6Network('::/128'),
        IPv6Network('::ffff:0:0/96'),
        IPv6Network('100::/64'),
        IPv6Network('2001::/23'),
        IPv6Network('2001:2::/48'),
        IPv6Network('2001:db8::/32'),
        IPv6Network('2001:10::/28'),
        IPv6Network('fc00::/7'),
        IPv6Network('fe80::/10'),
        ]

    _reserved_networks = [
        IPv6Network('::/8'), IPv6Network('100::/8'),
        IPv6Network('200::/7'), IPv6Network('400::/6'),
        IPv6Network('800::/5'), IPv6Network('1000::/4'),
        IPv6Network('4000::/3'), IPv6Network('6000::/3'),
        IPv6Network('8000::/3'), IPv6Network('A000::/3'),
        IPv6Network('C000::/3'), IPv6Network('E000::/4'),
        IPv6Network('F000::/5'), IPv6Network('F800::/6'),
        IPv6Network('FE00::/9'),
    ]

    _sitelocal_network = IPv6Network('fec0::/10')
```

return42 replied:
No, options for `_private_networks` and  `_reserved_networks` I would add in a follow up PR, I just linked them all to show, that there other properties to check [network.is_private](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_private) and [netwirk.is_reserved](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_reserved) ...

unixfox replied:
Ok then we can tackle this issue in another PR.

unixfox commented on searx/botdetection/_helpers.py:
```diff
@@ -63,7 +89,9 @@ def get_real_ip(request: flask.Request) -> str:
     real_ip = request.headers.get('X-Real-IP')
     remote_addr = request.remote_addr
-    logger.debug("X-Forwarded-For: %s || X-Real-IP: %s || request.remote_addr: %s", forwarded_for, real_ip, remote_addr)
+    # logger.debug(
```
Are those comments for debug or is it going to be implemented in the final code?

return42 replied:
I want to have this comments in the code .. its just for convenience to activate additional debug messages on my instance / I will do a lot of tests in the next weeks (also after a merge), sometimes it can be helpful to activate these additional (and very verbose) debug messages.  I will remove these comments from the code at some point.

SuperSandro2000 commented on searx/botdetection/limiter.toml:
```diff
@@ -0,0 +1,22 @@
+# to determine whether or not an address is part of a (client) network.
+
+ipv4_prefix = 32
+ipv6_prefix = 48
```
I have some doubt that this ipv6 prefix is generally correct and that we can find a general correct prefix. My provider for example hands me out /56

unixfox replied:
You can get /48 on some providers so that's why we consider this range to be a single network.

return42 replied:
> and that we can find a general correct prefix. 

@SuperSandro2000: that is true, but I would prefer /48 prefix over (e.g.) /56 .. there are two use cases we need to consider:

1. A bot in a /48 subnet where a user like you with a /56 prefix is also in (which will probably very rarely be the case).  --> the /48 subnet (and the bot) gets unblocked as soon a /56 user opens SearXNG in the WEB-Browser

2. A bot that owns a complete /48 subnet --> bot is permanently blocked

If we would use a /56 prefix, the bot could run through all its addresses in use case 2. and this would be in my opinion the much worse choice compared to a /48 prefix as described in use case 1.
